File: Example-Mac-Storyboard/Example-Mac-Storyboard/ViewController.m
//
//  ViewController.m
//  Example-Mac-Storyboard
//
//  Created by Stuart Tevendale on 08/11/2023.
//

#import "ViewController.h"
#import "PlaybackManager.h"
#import "DetailViewController.h"
#import "DetailWindowController.h"

#import <CocoaUPnP/CocoaUPnP.h>


@interface ViewController () <UPPDiscoveryDelegate, NSTableViewDelegate, NSTableViewDataSource>
@property (strong, nonatomic) NSTimer *searchTimer;
@property (strong, nonatomic) NSMutableArray *devices;
@property (strong, nonatomic) NSString *selectedInterface;
@property (strong, nonatomic) PlaybackManager *playbackManager;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    NSDictionary *interfaces = [SSDPServiceBrowser availableNetworkInterfaces];
    NSArray *interfaceNameArray = [interfaces allKeys];
    [networksPopup removeAllItems];
    
    // TODO: Set first item as currently selected interface
    for (NSString *interfaceName in interfaceNameArray) {
        NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:interfaceName action:@selector(popupDataTypeChanged:) keyEquivalent:@""];
        [menuItem setEnabled:YES];
        // Addition needed for Mac OS X 10.10 otherwise the menu items aren't enabled - worked on 10.9 without this
        [menuItem setTarget:self];
        [networksPopup.menu addItem:menuItem];
    }
    
    [[UPPDiscovery sharedInstance] addBrowserObserver:self];
    [self setupSearchTimer];
}

- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];

    // Update the view, if already loaded.
}


- (void)setupSearchTimer
{
    self.searchTimer = [NSTimer scheduledTimerWithTimeInterval:5.0
                                                        target:self
                                                      selector:@selector(timerFired:)
                                                      userInfo:nil
                                                       repeats:YES];
    [self.searchTimer fire];
}

- (void)timerFired:(NSTimer *)timer
{
//    [[UPPDiscovery sharedInstance] startBrowsingForServices:@"ssdp:all" withInterface:_selectedInterface];
    
    // Look for media servers
    [[UPPDiscovery sharedInstance] startBrowsingForServices:@"urn:schemas-upnp-org:service:ContentDirectory:1" withInterface:_selectedInterface];
}

- (IBAction)popupDataTypeChanged:(id)sender {
    NSMenuItem *item = (NSMenuItem *)sender;
    NSString *selectedNetwork = item.title;
    NSLog(@"Popup Changed: %@", selectedNetwork);
    _selectedInterface = selectedNetwork;
    [[UPPDiscovery sharedInstance] forgetAllKnownDevices];
    [_devices removeAllObjects];
    [tableView reloadData];
    [self setupSearchTimer];
}



#pragma mark - UPPDiscoveryDelegate

- (void)discovery:(UPPDiscovery *)discovery didFindDevice:(UPPBasicDevice *)device
{
    NSLog(@"In - (void)discovery:(UPPDiscovery *)discovery didFindDevice:(UPPBasicDevice *)device");
    if ([self.devices containsObject:device]) {
        return;
    }

    [self.devices addObject:device];
    
    [tableView reloadData];
}

- (void)discovery:(UPPDiscovery *)discovery didRemoveDevice:(UPPBasicDevice *)device
{
    NSLog(@"In - (void)discovery:(UPPDiscovery *)discovery didRemoveDevice:(UPPBasicDevice *)device");
    if (![self.devices containsObject:device]) {
        return;
    }

    [self.devices removeObject:device];
    
    [tableView reloadData];
}

#pragma mark - Lazy Instantiation

- (NSMutableArray *)devices
{
    if (!_devices) {
        _devices = [NSMutableArray array];
    }

    return _devices;
}


#pragma  mark - NSTableViewDataSource methods

- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return [self.devices count];
}

- (NSView *)tableView:(NSTableView *)tableView
   viewForTableColumn:(NSTableColumn *)tableColumn
                  row:(NSInteger)row {
    
    NSTableCellView *cellView = [tableView makeViewWithIdentifier:[tableColumn identifier] owner:self];
    
    UPPBasicDevice *device = self.devices[row];
    
    NSString *cellText = [NSString stringWithFormat:@"%@ - %@", device.friendlyName , device.deviceType];
    
    [cellView.textField setStringValue:cellText];

    return cellView;
}

- (void)tableViewSelectionDidChange:(NSNotification *)aNotification {
    NSLog(@"Selection Changed...");
    NSLog(@"Selected Row: %ld", tableView.selectedRow);
    id selectedDevice = [_devices objectAtIndex:tableView.selectedRow];
    NSLog(@"Device: %@", selectedDevice);
    
    if ([selectedDevice isKindOfClass:[UPPMediaRendererDevice class]]) {
        self.playbackManager.renderer = selectedDevice;
    }

    else if ([selectedDevice isKindOfClass:[UPPMediaServerDevice class]]) {
        NSString *objectId;
        
        
        NSStoryboard *sb = [NSStoryboard storyboardWithName:@"Main" bundle:nil];
        
        DetailWindowController *detailWindowController = (DetailWindowController *)[sb instantiateControllerWithIdentifier:@"detailWindow"];
  
        detailWindowController.window.title = [selectedDevice friendlyName];
        
        
        detailWindowController.playbackManager = self.playbackManager;
        detailWindowController.device = selectedDevice;
        detailWindowController.objectId = objectId;

        [detailWindowController showWindow:self];

     }
}

@end

File: Example-Mac-Swift/Example-Mac-Swift/ViewController.swift
//
//  ViewController.swift
//  Example-Mac-Swift
//
//  Created by Stuart Tevendale on 09/11/2023.
//

import Cocoa



class ViewController: NSViewController, UPPDiscoveryDelegate, NSTableViewDelegate, NSTableViewDataSource  {

    @IBOutlet weak var tableView: NSTableView!
    @IBOutlet weak var networksPopup: NSPopUpButton!
    
    var searchTimer: Timer = Timer()
    var devices: [UPPBasicDevice] = []
    var selectedInterface = ""
    var playbackManager: PlaybackManager = PlaybackManager()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // TODO: Set first item as currently selected interface
        let interfaces: Dictionary = SSDPServiceBrowser.availableNetworkInterfaces()
        let interfaceNameArray = interfaces.keys
        networksPopup.removeAllItems()
        for name in interfaceNameArray {
            networksPopup.addItem(withTitle: name as! String)
        }
        networksPopup.action = #selector(popupDataTypeChanged (_:))
        networksPopup.target = self
        
        UPPDiscovery.sharedInstance().addBrowserObserver(self)
        setupSearchTimer()
        
    }

    override var representedObject: Any? {
        didSet {
        // Update the view, if already loaded.
        }
    }
    
   @IBAction func popupDataTypeChanged(_ sender: NSPopUpButton) {
       self.selectedInterface = sender.titleOfSelectedItem ?? ""
       
       print("Selection changed: %s", sender.titleOfSelectedItem as Any)
       
       UPPDiscovery.sharedInstance().forgetAllKnownDevices()
       devices.removeAll()
       tableView.reloadData()
       setupSearchTimer()
       
        
    }
    
    func setupSearchTimer() {
        self.searchTimer = Timer(timeInterval: 5.0, target: self, selector: #selector(timerFired(timer: )), userInfo: nil, repeats: true)
        
        self.searchTimer.fire()
    }
    
    @IBAction func timerFired(timer: Timer) {
        UPPDiscovery.sharedInstance().startBrowsing(forServices: "ssdp:all", withInterface: self.selectedInterface)
    }
    
    // MARK: - UPPDiscoveryDelegate
    func discovery(_ discovery: UPPDiscovery, didFind device: UPPBasicDevice) {
        print("Found device" + device.friendlyName)
        if self.devices.contains(device) {
            return
        }
        
        self.devices.append(device)
        tableView.reloadData()
    }
    
    func discovery(_ discovery: UPPDiscovery, didRemove device: UPPBasicDevice) {
        if !(self.devices.contains(device)) {
            return
        }
        
        if let i = self.devices.firstIndex(of: device) {
            self.devices.remove(at: i)
        }
        
        tableView.reloadData()
    }
    
    // MARK: - NSTableViewDataSource methods
    
    func numberOfRows(in tableView: NSTableView) -> Int {
        return self.devices.count
    }
    
    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        
        guard let cellView = tableView.makeView(withIdentifier:
                NSUserInterfaceItemIdentifier(rawValue: "AutomaticTableColumnIdentifier.0"),
                owner: self) as? NSTableCellView
            else {
                return nil
            }
         
        let device: UPPBasicDevice = self.devices[row]
        let cellText = device.friendlyName + " - " + device.deviceType
        
        cellView.textField?.stringValue = cellText
        return cellView
    }
    
    func tableViewSelectionDidChange(_ notification: Notification) {
        print ("Selection changed")
        let selectedDevice = self.devices[tableView.selectedRow];
        
        if selectedDevice.isKind(of: UPPMediaRendererDevice.self) {
            self.playbackManager.renderer = selectedDevice as? UPPMediaRendererDevice
            
        }
        else if selectedDevice.isKind(of: UPPMediaServerDevice.self) {

            let sb = NSStoryboard(name: "Main", bundle: nil)
            let detailWindowController:DetailWindowController = sb.instantiateController(withIdentifier: "detailWindow") as! DetailWindowController
            detailWindowController.window?.title = selectedDevice.friendlyName
            detailWindowController.playbackManager = self.playbackManager
            detailWindowController.device = selectedDevice as? UPPMediaServerDevice ?? nil
            
            detailWindowController.showWindow(self)
        }
    }

}

